# Examine variability of gene expression in SNs, DRG, GTEx

## Introduction

Here we examine the relative standard deviation (RSD, also known as coefficient of variation) of gene expression across samples for our sensory neuron cells as well as GTEx tissues, DRG, and IPSCs.
We set a threshold to define genes as expressed, since very lowly expressed genes will generally have higher relative variability across samples.

```{r Setup}
library(biomaRt) # biomaRt must be loaded before dplyr to avoid overriding 'select'
library(data.table)
library(dplyr)
library(tidyr)
library(GOsummaries)
library(ggplot2)
library(gridExtra)
options(stringsAsFactors=F)

# This file is generated by compare.SN.GTEx.Rmd
rpkmFname <- "results/MDS/rpkms.gtex.ips.drg.sn.rds"
rpkmFname <- "results/MDS/new.rpkms.gtex.ips.drg.sn.rds"
sampleMetaFname <- "results/MDS/rpkms.sample_meta.txt"
outputDir <- "results/expression_variance/rpkm1/"
rpkmThreshold <- 1
print(paste0("RPKM threshold for expressed genes: ", rpkmThreshold))
knitting = T
fontSize = 16
goSummariesFontSize = 12
```

Load RPKM files. For IPS-derived SNs exclude outlier samples and only use samples differentiated using the P2V2 protocol.

```{r LoadFiles}
rpkm.df <- readRDS(rpkmFname)
print("Number of genes with RPKM data in all datasets (HIPSCI IPS, SNs, DRG, GTEx) :")
dim(rpkm.df)
colnames(rpkm.df) = tolower(colnames(rpkm.df))
colnames(rpkm.df)[1] = "gene_id"
rownames(rpkm.df) = rpkm.df$gene_id

sample.meta <- read.delim(sampleMetaFname)
sample.meta$SAMPID = tolower(sample.meta$SAMPID)
# Shorten a couple cell type names
sample.meta$SMTSD[sample.meta$SMTSD == "Cells - EBV-transformed lymphocytes"] = "LCLs"
sample.meta$SMTSD[sample.meta$SMTSD == "Cells - Transformed fibroblasts"] = "Fibroblasts"

# Subset sample.meta to those samples we have RPKM data for
sample.meta = sample.meta %>% filter(SAMPID %in% colnames(rpkm.df))

readRPKMTable = function(fname) {
  # The first 2 columns are gene_id and length
  df = readr::read_tsv(fname) %>% as.data.frame()
  # Fix gene names to remove the version number (e.g. ".4" in ENSG00000223972.4)
  df$gene_id <- gsub("\\.[\\d]+", "", df$gene_id, perl=T)
  rownames(df) <- df$gene_id
  # include gene_id
  df[, c(1, 3:ncol(df))]
}
sn.df.all = readRPKMTable("../../sn_hg37/sn.hg37.gencode_full.rpkm.txt.gz")
colnames(sn.df.all) = tolower(colnames(sn.df.all))

sn.meta.all = read.delim("../data/metadata.all.txt")
sn.meta.all$Sample = tolower(sn.meta.all$Sample)
sn.meta.all = sn.meta.all %>% filter(! Sample %in% c("posc_1", "iakz_1", "koun_2", "yuze_1_1", "yuze_1_2"))
sn.df.all = sn.df.all %>% dplyr::select(-one_of(c("posc_1", "iakz_1", "koun_2", "yuze_1_1", "yuze_1_2")))

sn.meta.all$RIN[sn.meta.all$RIN == "X"] = 5
sn.meta.all$RIN = as.numeric(sn.meta.all$RIN)
# In one run of the expression variance code, we exclude samples with low (or unknown) RIN numbers
lowRINsamples = sn.meta.all$Sample[sn.meta.all$RIN < 8 | is.na(sn.meta.all$RIN)]
sn.meta = sn.meta.all
sn.df = sn.df.all

# Each of these 6 lines had 3 differentiation replicates (and no extraction replicates)
# All other differentiation replicates were across P1 vs P2V2 protocols, except for OOMZ
sn.meta.rep <- sn.meta %>% filter(donor %in% c("guss", "hehd", "nukw", "pelm", "podx", "qaqx"))
sn.meta.P2 <- sn.meta %>% filter(Protocol == "P2V2") %>% mutate(sampleDifferentiation=paste(Cell.line, Differentiaton.Replicate))
# Now remove extraction replicates, as these aren't interesting for
# investigating expression variability
sn.meta.P2 <- sn.meta.P2[!duplicated(sn.meta.P2$sampleDifferentiation),]
```

First determine variability across all IPS-derived SNs (P2V2 protocol only), measured as relative standard deviation (RSD) (coefficient of variation). We consider only genes with (mean RPKM across samples) > 1, so we don't just get highly variable genes that are very lowly expressed.

```{r CalculateRSDs}
trimmedMean = function(x) { mean(x, trim=0.1, na.rm=T) }
mean_na_rm = function(x) { mean(x, na.rm=T) }
# Add annotations to the genes
count.data = readRDS("../data/combined_expression_data.v5.rds")
annotation = count.data$gene_metadata

# Add Hipsci and DRG data to the annotation
#hipsci.rpkm.df = read.delim("../../reference/hipsci/hipsci.REL-2016-09.rpkm.txt.gz")
hipsci.rpkm.df <- readRPKMTable("../../hipsci/hipsci.hg37.gencode_full.rpkm.txt.gz")
hipsci.rpkm.df = hipsci.rpkm.df[, -1]
hipsci.rpkmavg.df = data.frame(gene_id = rownames(hipsci.rpkm.df), hipsci_mean_rpkm=rowMeans(hipsci.rpkm.df))
rm(hipsci.rpkm.df)

drg.rpkm.df = readRPKMTable("../../drg/drg.hg37.gencode_full.rpkm.txt.gz")
drg.meta = readr::read_csv("../../drg/drg.sample_data.csv")
# Exclude the DRG outlier sample
drg.meta = drg.meta %>% filter(Sample != "T32053")
drg.rpkm.df = drg.rpkm.df[, as.character(drg.meta$Sample)]
drg.rpkmavg.df = data.frame(gene_id = rownames(drg.rpkm.df), drg_mean_rpkm=rowMeans(drg.rpkm.df))

# Add single cell data about genes to the annotation
sc.diffexp.df <- fread("results/celltypes/singleCell.differentialExpression.txt")
sc.diffexp.df <- subset(sc.diffexp.df, select=-c(lfcSE, stat, pvalue))

# Add OpenTargets pain-associated gene information
cttv.df <- fread("../data/cttv/targets_associated_with_pain.2016_8_22.txt")

annotation %<>% left_join(hipsci.rpkmavg.df, by="gene_id") %>%
  left_join(drg.rpkmavg.df, by="gene_id") %>%
  left_join(sc.diffexp.df, by=c("gene_id" = "gene")) %>%
  left_join(cttv.df, by=c("gene_name" = "CTTV_symbol")) %>%
  rename(sc.baseMean = baseMean, sc.log2FoldChange = log2FoldChange)

rpkm.mat.sn = as.matrix(sn.df[,sn.meta.P2$Sample])
rownames(rpkm.mat.sn) = sn.df$gene_id
getCVSummary = function(df) {
  df.summary = data.frame(gene_id = rownames(df), avgrpkm = apply(df, 1, FUN=mean_na_rm))
  df.summary$logavgrpkm = log(df.summary$avgrpkm)
  df.summary$rsd = apply(df, 1, FUN=function(x) {sd(x, na.rm=T) / mean_na_rm(x)})
  df.summary = df.summary[order(-df.summary$rsd), ]
  numExpressed = sum(df.summary$avgrpkm > rpkmThreshold)
  print(paste("Number of genes with mean RPKM across samples > ", rpkmThreshold, ":", numExpressed))
  df.summary
}
rpkm.summary.sn = getCVSummary(rpkm.mat.sn)
rpkm.summary.sn$cell_type = "IPSDSN"
ggplot(rpkm.summary.sn, aes(logavgrpkm)) +
  geom_histogram(binwidth = 0.2) +
  theme_gray(fontSize) +
  ggtitle("Sensory neuron gene RPKM")

rpkm.summary.sn.x = rpkm.summary.sn[rpkm.summary.sn$avgrpkm > rpkmThreshold,]
ggplot(rpkm.summary.sn.x, aes(log(rsd))) +
  geom_histogram() +
  theme_gray(fontSize) +
  ggtitle("Sensory neuron gene RSD")

# Save a table with the gene RSDs
write.table(rpkm.summary.sn.x, paste0(outputDir, "sn.gene.rsd.txt"), quote=F, row.names=F, sep="\t")
```

We can split genes down by biotype.

```{r RSD_Gene_biotype}
rpkm.summary.sn.ann = rpkm.summary.sn %>%
  left_join(annotation, by="gene_id")
  
rpkm.summary.sn.x.ann = rpkm.summary.sn.x %>%
  left_join(annotation, by="gene_id")

fname = paste0(outputDir, "rpkm.summary.sn.txt")
write.table(rpkm.summary.sn.ann, fname, quote=F, sep="\t", row.names=F, col.names=T)
fname = paste0(outputDir, "rpkm.summary.sn.expressed.txt")
write.table(rpkm.summary.sn.x.ann, fname, quote=F, sep="\t", row.names=F, col.names=T)

rpkm.summary.sn.ann = rpkm.summary.sn.ann %>%
  filter(gene_biotype %in% c("protein_coding", "lincRNA", "antisense", "processed_transcript"))

rpkm.summary.sn.x.ann = rpkm.summary.sn.x.ann %>%
  filter(gene_biotype %in% c("protein_coding", "lincRNA", "antisense", "processed_transcript"))

ggplot(rpkm.summary.sn.x.ann, aes(logavgrpkm, fill=gene_biotype)) +
  geom_density(alpha = 0.5) +
  theme_gray(fontSize) +
  ggtitle("Sensory neuron RPKM by gene biotype")

ggplot(rpkm.summary.sn.x.ann[rpkm.summary.sn.x.ann$rsd <= 2,], aes(log(rsd), fill=gene_biotype)) +
  geom_density(alpha = 0.5) +
  theme_gray(fontSize) +
  ggtitle("Sensory neuron RSD by gene biotype")

table(rpkm.summary.sn.x.ann$gene_biotype) / sum(table(rpkm.summary.sn.x.ann$gene_id))
```

This shows that protein-coding genes are more highly expressed, and on average have slightly lower RSD, though some protein-coding genes have high RSD. However, note that the great majority of genes we're considering are protein-coding (>96%), since we set an RPKM threshold.
Next we see if the highly (or lowly) variable genes are enriched in certain pathways/processes.
We also want to get fold enrichments for each category. GoSummaries / gprofiler don't give us this by default. I spent a while trying to get info from the gProfiler web site, and then also from the Panther GO enrichment web site. I figured out how Panther calculates fold enrichment -- first the expected number of overlaps with a term is determined: expoverlaps = query size * (term size / total genome genes). Then it’s just actual overlaps divided by expoverlaps. But the “genome size” isn’t specified anywhere in gprofiler. Assuming 20,000 genes seems reasonable enough.

```{r GOenrichment}
# Here we just look at the top 1000 genes by RSD
rpkm.summary.sn.x = rpkm.summary.sn.x %>% arrange(-rsd)
fname = paste0(outputDir, "IPSDSN.highRSDgenes.forGOanalysis.txt")
write.table(rpkm.summary.sn.x %>% dplyr::select(gene_id, rsd) %>% .[1:1000,], fname, quote=F, sep="\t", row.names=F)
# Now use GeneTrail2 to test for gene set over-representation

gl = list(rpkm.summary.sn.x[1:1000,]$gene_id)
gs = gosummaries(gl, max_signif=100)
plot(gs, fontsize = goSummariesFontSize)
gs.results.df = attributes(gs)$gprofiler_results

totalGenes = 2e4
gs.results.df$expectedOverlaps = gs.results.df$query.size * gs.results.df$term.size / totalGenes
gs.results.df$foldEnrichment = gs.results.df$overlap.size / gs.results.df$expectedOverlaps

fname = paste0(outputDir, "IPSDSN.1000.highRSDgenes.gosummaries.results.txt")
write.table(gs.results.df[,4:ncol(gs.results.df)], fname, quote=F, sep="\t", row.names=F)

# Also write out the full set of expressed genes, which can be
# used in gProfiler as a background set
fname = paste0(outputDir, "expressedGenes.IPSDSN.txt")
write.table(rpkm.summary.sn.x$gene_id, fname, quote=F, sep="\t", row.names=F, col.names=F)

# Here we look at the top 500 genes by RSD vs. the bottom 500
N = nrow(rpkm.summary.sn.x)
highRSD = rpkm.summary.sn.x[1:500,]$gene_id
lowRSD = rev(rpkm.summary.sn.x[(N-499):N,]$gene_id)
gl2 = list(List=list(highRSD, lowRSD))
gs2 = gosummaries(gl2)
plot(gs2, fontsize = goSummariesFontSize)
```

We see that the top 1000 variable genes are enriched in similar processes as we see in the principal components of gene expression (developmental and pluripotency genes as well as extracellular matrix). This is basically the same if we contrast the top 500 vs. bottom 500 genes by RSD.

As a quality check, let's see if the inter-sample variability is different if we consider IPSDSN samples with high neuron content or high fibroblast content only.

```{r CalculateRSD_SN_Subsets}
getRSDSummary = function(rpkm.mat, cell_type) {
  rpkm.summary = data.frame(gene_id = rpkm.df.sn$gene_id, avgrpkm = apply(rpkm.mat, 1, FUN=mean_na_rm))
  rpkm.summary$logavgrpkm = log(rpkm.summary$avgrpkm)
  ggplot(rpkm.summary, aes(logavgrpkm)) +
    geom_histogram(binwidth = 0.2) +
    theme_gray(fontSize) +
    ggtitle(paste(cell_type, " gene RPKM"))
  numExpressed = sum(rpkm.summary$avgrpkm > rpkmThreshold)
  print(paste("Number of genes with mean RPKM across samples > ", rpkmThreshold, ":", numExpressed))
  rpkm.summary$cell_type = cell_type
  rpkm.summary$rsd = apply(rpkm.mat, 1, FUN=function(x) {sd(x, na.rm=T) / mean_na_rm(x)})
  rpkm.summary = rpkm.summary[order(-rpkm.summary$rsd), ]
  rpkm.summary.x = rpkm.summary[rpkm.summary$avgrpkm > rpkmThreshold,]
  rpkm.summary.x
}

quantile(sn.meta.P2$fibroblast_fraction.V5, probs=seq(0, 1, 0.01))

rpkm.mat.sn_low = as.matrix(rpkm.df.sn[ ,sn.meta.P2[sn.meta.P2$fibroblast_fraction.V5 < 0.20, ]$Sample])
rownames(rpkm.mat.sn_low) = rpkm.df.sn$gene_id
rpkm.summary.sn_low.x = getRSDSummary(rpkm.mat.sn_low, "IPSDSN-low")

rpkm.mat.sn_med = as.matrix(rpkm.df.sn[ ,sn.meta.P2[sn.meta.P2$fibroblast_fraction.V5 >= 0.2 & sn.meta.P2$fibroblast_fraction.V5 <= 0.5, ]$Sample])
rownames(rpkm.mat.sn_med) = rpkm.df.sn$gene_id
rpkm.summary.sn_med.x = getRSDSummary(rpkm.mat.sn_med, "IPSDSN-med")

rpkm.mat.sn_hi = as.matrix(rpkm.df.sn[ ,sn.meta.P2[sn.meta.P2$fibroblast_fraction.V5 > 0.5, ]$Sample])
rownames(rpkm.mat.sn_hi) = rpkm.df.sn$gene_id
rpkm.summary.sn_hi.x = getRSDSummary(rpkm.mat.sn_hi, "IPSDSN-hi")
dim(rpkm.mat.sn_low)
dim(rpkm.mat.sn_med)
dim(rpkm.mat.sn_hi)

rpkm.summary.sn_subsets = rbind(rpkm.summary.sn_low.x, rpkm.summary.sn_med.x, rpkm.summary.sn_hi.x)

p = ggplot(rpkm.summary.sn_subsets, aes(log(rsd), fill=cell_type)) + geom_density(alpha = 0.5) + theme_bw(fontSize)
print(p)

# Look at GO enrichment of the top 1000 genes by RSD in each sample subset
writeGosummariesFile = function(rpkm.summary.df, fnamePart) {
  rpkm.summary.df = rpkm.summary.df %>% arrange(-rsd)
  gl = list(rpkm.summary.df[1:1000,]$gene_id)
  gs = gosummaries(gl, max_signif=100)
  plot(gs, fontsize = goSummariesFontSize)
  gs.results.df = attributes(gs)$gprofiler_results
  totalGenes = 2e4
  gs.results.df$expectedOverlaps = gs.results.df$query.size * gs.results.df$term.size / totalGenes
  gs.results.df$foldEnrichment = gs.results.df$overlap.size / gs.results.df$expectedOverlaps
  fname = paste0(outputDir, "IPSDSN.highRSDgenes.gosummaries.", fnamePart, ".txt")
  write.table(gs.results.df[,4:ncol(gs.results.df)], fname, quote=F, sep="\t", row.names=F)
  
  fname = paste0(outputDir, "IPSDSN.rsd.summary.", fnamePart, ".txt")
  write.table(rpkm.summary.df, fname, quote=F, sep="\t", row.names=F)
}
writeGosummariesFile(rpkm.summary.sn_low.x, "lowfib")
writeGosummariesFile(rpkm.summary.sn_med.x, "medfib")
writeGosummariesFile(rpkm.summary.sn_hi.x, "highfib")
```

RSDs are slightly lower (left-shifted distribution) in IPSDSNs estimated to have low fibroblast content (< 20%), intermediate in those with intermediate fibroblast content (40-60%) and slightly higher in those with high fibroblast content (> 80%). The full set of samples is somewhat intermediate between these, but more similar to the full set with high fibroblast content.
Overall, this tells me that the more pure SN samples had slightly lower RSDs across genes - i.e. the fibroblast content IS a contributing factor to gene expression variability across samples.

Let's see if the fact that samples are repeated from the same DRG donors affect the computed CV distribution.

```{r CalculateRSDs.DRG}
# Get a summary of the CV just across the 10 unique DRG donors
drg.unique.samples = as.character(drg.meta[drg.meta$Selected == 1, ]$Sample)
drg.rpkm.10 = drg.rpkm.df[, drg.unique.samples]
drg.10.summary = getCVSummary(drg.rpkm.10)
drg.10.summary$cell_type = "DRG_10"
drg.10.summary.x = drg.10.summary[drg.10.summary$avgrpkm > rpkmThreshold,]

# DRG seems to have more expressed genes at RPKM > 1, which have lower
# median expression than genes with RPKM > 1 in SNs, IPSCs, etc.
# If we want to make the average expression level of the genes we consider
# equivalent, we would do this.
#drg.10.summary.x = drg.10.summary.x[drg.10.summary.x$avgrpkm > 2.1,]

# Do the same for the 28 DRG samples
drg.summary = getCVSummary(drg.rpkm.df)
drg.summary$cell_type = "DRG_all"
drg.summary.x = drg.summary[drg.summary$avgrpkm > rpkmThreshold,]

drg.subsets.summary = rbind(drg.summary.x, drg.10.summary.x)
# Export at size 6x2.8
ggplot(drg.subsets.summary, aes(log(rsd), fill=cell_type)) +
  geom_density(alpha = 0.5) + theme_bw(12) +
  xlab("log(CV)") + scale_fill_discrete(name="Samples")

fname = paste0(outputDir, "DRG.gene.cv.summary.txt")
write.table(drg.summary.x, fname, quote=F, sep="\t", row.names=F)
```

Fantastic - there is no reduction in CV distribution given that many of the samples are from the same donors. That is, gene expression among the 10 DRGs from unique donors is the same as that among the 28 samples (which include multiple samples from given donors).

Let's compare the distribution of RSD between the sensory neurons and GTEx tissue samples.

```{r CalculateRSDsOther}
calculateRSDs = function(sampleIDs, cellType) {
  rpkm.df.sample = rpkm.df[,c("gene_id", sampleIDs)] %>% filter(gene_id %in% annotation$gene_id)
  rpkm.mat = as.matrix(rpkm.df.sample[,sampleIDs])
  rownames(rpkm.mat) = rpkm.df.sample$gene_id

  rpkm.summary = data.frame(gene_id = rpkm.df.sample$gene_id, avgrpkm = apply(rpkm.mat, 1, FUN=mean_na_rm))
  rpkm.summary$logavgrpkm = log(rpkm.summary$avgrpkm)
  numExpressed = sum(rpkm.summary$avgrpkm > rpkmThreshold)
  print(paste("Number of genes with mean RPKM across samples > ", rpkmThreshold, ":", numExpressed))
  rpkm.summary$cell_type = cellType
  rpkm.summary$rsd = apply(rpkm.mat, 1, FUN=function(x) {sd(x, na.rm=T) / mean(x, na.rm=T)})
  rpkm.summary
}

plotTissueRSDs = function(tissueNames, makePlots=F, subsample=1, topN=0) {
  rpkm.summary.df = rpkm.summary.sn.x
  if (topN > 0) {
    rpkm.summary.df = rpkm.summary.df[order(rpkm.summary.df$avgrpkm, decreasing=T),][1:topN, ]
  }
  for (tissue in tissueNames) {
    print(tissue)
    sampleIDs = sample.meta$SAMPID[sample.meta$SMTSD == tissue]
    if (subsample < 1) {
      sampleIDs = sample(sampleIDs, size=length(sampleIDs) * subsample, replace=F)
    }
    rpkm.summary.tissue = calculateRSDs(sampleIDs, tissue)
    
    # If topN is defined then we use this number of top genes, so that the
    # same number of genes is used in each tissue. Otherwise we use the global
    # RPKM threshold.
    if (topN > 0) {
      rpkm.summary.tissue.x = rpkm.summary.tissue[order(rpkm.summary.tissue$avgrpkm, decreasing=T),][1:topN, ]
      print(paste("Mean/median RPKM among top", topN, "genes:", mean(rpkm.summary.tissue.x$avgrpkm), "/", median(rpkm.summary.tissue.x$avgrpkm)))
    } else {
      rpkm.summary.tissue.x = rpkm.summary.tissue[rpkm.summary.tissue$avgrpkm > rpkmThreshold,]
      print(paste("Mean/median RPKM among genes above threshold:", mean(rpkm.summary.tissue.x$avgrpkm), "/", median(rpkm.summary.tissue.x$avgrpkm)))
    }
    rpkm.summary.df = rbind(rpkm.summary.df, rpkm.summary.tissue.x)
    if (makePlots) {
      p = ggplot(rpkm.summary.df, aes(log(rsd), fill=cell_type)) + geom_density(alpha = 0.5) + theme_bw(fontSize)
      print(p)
    }
  }
  rpkm.summary.df
}

tissueNames = names(table(sample.meta$SMTSD)[table(sample.meta$SMTSD) > 20])
tissueNames = tissueNames[!(tissueNames %in% c("Sensory neuron P1", "Sensory neuron P2"))]
selectTissues = c("DRG", "IPSC", "Nerve - Tibial", "Brain - Cortex", "Brain - Hypothalamus", "LCLs", "Fibroblasts", "Whole Blood")
#rpkm.summary.df = plotTissueRSDs(selectTissues, makePlots=T)

# Get RSD summaries for all tissues
rpkm.summary.rpkm_gt.df = plotTissueRSDs(tissueNames, makePlots=F)
rpkm.summary.df = rpkm.summary.rpkm_gt.df

gzf = gzfile(paste0(outputDir, "rpkm.summary.df.txt.gz"), "w")
write.table(rpkm.summary.df, gzf, quote=F, sep="\t", col.names=T, row.names=F)
close(gzf)

# Set only a couple of the cell types to be coloured
tn = factor(c(tissueNames[!tissueNames %in% c("Whole Blood")], "Whole Blood", "IPSDSN"), levels = 
c(tissueNames[!tissueNames %in% c("Whole Blood")], "Whole Blood", "IPSDSN"))
rpkm.summary.df$cell_type = factor(rpkm.summary.df$cell_type, levels=levels(tn))
p = ggplot(rpkm.summary.df, aes(log(rsd), fill=cell_type)) + geom_density() + theme_bw(fontSize)
p = p + scale_fill_manual(breaks = tn,
                          values=c(rep(rgb(0.5, 0.5, 0.5, alpha=0.5), length(tissueNames)-1), rgb(0, 0, 1, alpha=0.5), rgb(1, 0, 0, alpha=0.5))) + theme(legend.position="none") + xlab("log(CV)")
print(p)

# Do the same but also include DRG and IPSC
rpkm.summary.df2 = rpkm.summary.df
# Set only a couple of the cell types to be coloured
highlightedTissues = c("Whole Blood", "IPSC", "DRG", "IPSDSN")
tn = factor(c(tissueNames[!tissueNames %in% highlightedTissues], highlightedTissues),
            levels = c(tissueNames[!tissueNames %in% highlightedTissues], highlightedTissues))
rpkm.summary.df2$cell_type = factor(as.character(rpkm.summary.df2$cell_type), levels=levels(tn))
p = ggplot(rpkm.summary.df2, aes(log10(rsd), fill=cell_type)) + geom_density() + theme_bw(fontSize)
p = p + scale_fill_manual(breaks = tn,
                          values=c(rep(rgb(0.5, 0.5, 0.5, alpha=0.5), length(tissueNames)-3), rgb(0, 0, 1, alpha=0.5), rgb(0.1, 0.7, 0.1, alpha=0.5), rgb(1, 0.65, 0, alpha=0.5), rgb(1, 0, 0, alpha=0.7))) +
  theme(legend.position="none") + xlab(expression(log[10](CV)))
#print(p)
print(p + xlim(c(-1.25, 1.05)))

# This next plot is just to get the legend, which merges "other" GTEx tissues together,
# and which we'll paste onto the figure just above
rpkm.summary.df2 = rpkm.summary.df
rpkm.summary.df2$cell_type = as.character(rpkm.summary.df2$cell_type)
#rpkm.summary.df2$cell_type[!(rpkm.summary.df$cell_type %in% c("Whole Blood", "IPSDSN"))] = "GTEx-other tissues"
rpkm.summary.df2$cell_type[!(rpkm.summary.df$cell_type %in% highlightedTissues)] = "GTEx-other tissues"
rpkm.summary.df2$cell_type = factor(rpkm.summary.df2$cell_type, levels=c("GTEx-other tissues", highlightedTissues))
p = ggplot(rpkm.summary.df2, aes(log10(rsd), fill=cell_type)) + geom_density() + theme_bw(fontSize) + theme(legend.title=element_blank())
p = p + scale_fill_manual(breaks = levels(rpkm.summary.df2$cell_type),
                          values=c(rgb(0.5, 0.5, 0.5, alpha=0.5), rgb(0, 0, 1, alpha=0.5), rgb(0.1, 0.7, 0.1, alpha=0.5), rgb(1, 0.65, 0, alpha=0.5), rgb(1, 0, 0, alpha=0.7))) +
  xlab(expression(log[10](CV)))
print(p)


# Get quantiles of RSD for each tissue
rsd.quantiles.df = data.frame()
for (tissue in c("IPSDSN", tissueNames)) {
  q = quantile(rpkm.summary.df$rsd[rpkm.summary.df$cell_type == tissue], probs = c(0.5, 0.9, 0.95, 0.99), na.rm=T)
  avg = mean(rpkm.summary.df$rsd[rpkm.summary.df$cell_type == tissue])
  rsd.quantiles.df = rbind(rsd.quantiles.df, c(cell_type = tissue, avg, q))
}
colnames(rsd.quantiles.df) = c("cell_type", "mean", "50%", "90%", "95%", "99%")

qprobs = c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99, 0.995, 0.999)
q = quantile(rpkm.summary.df %>% filter(cell_type != "IPSDSN") %>% .[["rsd"]], probs = qprobs, na.rm=T)
q
# 99% of genes in tissue samples have RSD < 1.57, but this varies by tissue
sum((rpkm.summary.df %>% filter(cell_type == "IPSDSN") %>% .[["rsd"]]) > q["99%"])
sum((rpkm.summary.df %>% filter(cell_type == "DRG") %>% .[["rsd"]]) > q["99%"])
sum((rpkm.summary.df %>% filter(cell_type == "Nerve - Tibial") %>% .[["rsd"]]) > q["99%"])
sum((rpkm.summary.df %>% filter(cell_type == "Brain - Hypothalamus") %>% .[["rsd"]]) > q["99%"])
sum((rpkm.summary.df %>% filter(cell_type == "Whole Blood") %>% .[["rsd"]]) > q["99%"])

# Save a bunch of memory!
rm(rpkm.df)
gc()
```

Across tissues, 99% of genes in tissue samples have RSD < 1.57, but this varies by tissue. The SNs have genes with higher variability across samples than do most tissue samples, but a few tissues are outliers, such as whole blood.
Let's see how RSD in our neurons compares with DRG.

```{r RSD_compare_DRG}
median(rpkm.summary.df[rpkm.summary.df$cell_type == "DRG",]$rsd)
median(rpkm.summary.df[rpkm.summary.df$cell_type == "IPSDSN",]$rsd)

quantile(rpkm.summary.df[rpkm.summary.df$cell_type == "DRG",]$rsd, probs=c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99))
quantile(rpkm.summary.df[rpkm.summary.df$cell_type == "IPSDSN",]$rsd, probs=c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99))
```

Let's see whether RSD is affected by whether the SN samples were received on feeders or E8.

```{r CalculateRSD_SN_Subsets.2}
table(sn.meta.P2$wasFeeder)

rpkm.mat.feeder = as.matrix(rpkm.df.sn[ , sn.meta.P2[sn.meta.P2$wasFeeder == "YES", ]$Sample])
rownames(rpkm.mat.feeder) = rpkm.df.sn$gene_id
rpkm.summary.feeder.x = getRSDSummary(rpkm.mat.feeder, "IPSDSN-feeder")

rpkm.mat.E8 = as.matrix(rpkm.df.sn[ , sn.meta.P2[sn.meta.P2$wasFeeder != "YES", ]$Sample])
rownames(rpkm.mat.E8) = rpkm.df.sn$gene_id
rpkm.summary.E8.x = getRSDSummary(rpkm.mat.E8, "IPSDSN-E8")

rpkm.mat.all = as.matrix(rpkm.df.sn[, sn.meta.P2$Sample])
rpkm.summary.sn.x.new = getRSDSummary(rpkm.mat.all, "IPSDSN-all")

rpkm.summary.sn_subsets.feeder = rbind(rpkm.summary.feeder.x, rpkm.summary.E8.x, rpkm.summary.sn.x.new)
p = ggplot(rpkm.summary.sn_subsets.feeder, aes(log(rsd), fill=cell_type)) +
  geom_density(alpha = 0.5) + theme_bw(fontSize) + xlab("log(CV)")
print(p)
```

The E8 samples seem to have slightly lower RSD globally than feeder samples. Let's see how this compares with the RSD distribution in IPSCs and DRG.

```{r CalculateRSD_SN_Subsets.3}
rpkm.summary.ips.drg = rpkm.summary.df %>% filter(cell_type %in% c("DRG", "IPSC"))
rpkm.summary.feeder_subsets = rbind(rpkm.summary.feeder.x, rpkm.summary.E8.x, rpkm.summary.ips.drg)

highlightedTissues = c("DRG", "IPSC", "IPSDSN-E8", "IPSDSN-feeder")
rpkm.summary.feeder_subsets$cell_type = factor(as.character(rpkm.summary.feeder_subsets$cell_type), levels=highlightedTissues)

p = ggplot(rpkm.summary.feeder_subsets, aes(log(rsd), fill=cell_type)) +
  geom_density(alpha = 0.6) + theme_bw(11) + xlab("log(CV)")
# Manually specify colors to be similar to previous (all tissues) density plot
p = p + scale_fill_manual(breaks = levels(rpkm.summary.feeder_subsets$cell_type),
                          values=c(rgb(1, 0.7, 0), rgb(0.1, 0.6, 0.1), rgb(1, 0, 0.4), rgb(1, 0.4, 0))) + xlab("log(CV)")
# Export at size 
print(p)
```

Looking again at how fibroblast content associates with RSD, this time comparing to DRG, and doing it among only E8-IPSDSN samples.

```{r CalculateRSD_SN_Subsets2}
sn.meta.P2.E8 = sn.meta.P2 %>% filter(wasFeeder == "NO")
rpkm.mat.sn_low = as.matrix(rpkm.df.sn[ ,sn.meta.P2.E8[sn.meta.P2.E8$fibroblast_fraction.V5 < 0.20, ]$Sample])
rownames(rpkm.mat.sn_low) = rpkm.df.sn$gene_id
rpkm.summary.sn_low.x = getRSDSummary(rpkm.mat.sn_low, "IPSDSN-low")

rpkm.mat.sn_med = as.matrix(rpkm.df.sn[ ,sn.meta.P2.E8[sn.meta.P2.E8$fibroblast_fraction.V5 >= 0.2 & sn.meta.P2.E8$fibroblast_fraction.V5 <= 0.45, ]$Sample])
rownames(rpkm.mat.sn_med) = rpkm.df.sn$gene_id
rpkm.summary.sn_med.x = getRSDSummary(rpkm.mat.sn_med, "IPSDSN-med")

rpkm.mat.sn_hi = as.matrix(rpkm.df.sn[ ,sn.meta.P2.E8[sn.meta.P2.E8$fibroblast_fraction.V5 > 0.45, ]$Sample])
rownames(rpkm.mat.sn_hi) = rpkm.df.sn$gene_id
rpkm.summary.sn_hi.x = getRSDSummary(rpkm.mat.sn_hi, "IPSDSN-hi")
dim(rpkm.mat.sn_low)
dim(rpkm.mat.sn_med)
dim(rpkm.mat.sn_hi)

rpkm.summary.sn_subsets = rbind(rpkm.summary.sn_low.x, rpkm.summary.sn_med.x, rpkm.summary.sn_hi.x)

rpkm.summary.sn_subsets = rbind(rpkm.summary.sn_subsets, rpkm.summary.df[rpkm.summary.df$cell_type == "DRG",])
highlightedTissues = c("DRG", "IPSDSN-low", "IPSDSN-med", "IPSDSN-hi")
rpkm.summary.sn_subsets$cell_type = factor(as.character(rpkm.summary.sn_subsets$cell_type), levels=highlightedTissues)

p = ggplot(rpkm.summary.sn_subsets, aes(log(rsd), fill=cell_type)) +
  geom_density(alpha = 0.6) + theme_bw(11) + xlab("log(CV)")
# Manually specify colors to be similar to previous (all tissues) density plot
p = p + scale_fill_manual(breaks = levels(rpkm.summary.sn_subsets$cell_type),
                          values=c(rgb(1, 0.7, 0), rgb(0.1, 0.7, 0.1), rgb(0.8, 0, 0.7), rgb(1, 0.4, 0))) + xlab("log(CV)")
# Export at 6x3
print(p)
```

Let's look at the average RSD of genes in certain GO terms of interest, relative to all expressed genes. We consider only genes above an expression threshold (e.g. RPKM > 1) since lowly expressed genes have higher RSD.

```{r GOterms}
rpkm.threshold = 1

sn.rsd.df = rpkm.summary.sn[, c("gene_id", "rsd", "avgrpkm")]
sn.go.df = sn.rsd.df
sn.go.df$gene_set = "IPSDSN-All genes"

terms = c("axonogenesis", "cell_differentiation", "cell_cycle",
          "CNS_development", "extracell_matrix_org",
          "nervous_system_dev", "neurogenesis",
          "sensory_organ_dev", "synapse_organization", "synaptic_signaling")
#terms = c("axonogenesis")

for (term in terms) {
  term.df = read.delim(paste0(outputDir, "/go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)
  term.df$gene_set = paste0("IPSDSN-", term)
  term.df = term.df %>% dplyr::left_join(sn.rsd.df) %>% na.omit()
  go.df2 = rbind(sn.go.df, term.df)
  go.df2.x = go.df2 %>% filter(avgrpkm > rpkm.threshold)
  p = ggplot(go.df2.x, aes(log(rsd), fill=gene_set)) +
    geom_density(alpha = 0.5) + theme_bw(fontSize) +
    xlab("log(CV)")
  print(p)
}
```

Many GO terms associated with nervous system development and differentiation show higher variability than typical genes.
Do the same thing considering only low fibroblast content samples.

```{r GOterms2}
rpkm.threshold = 1
sn.rsd.df = rpkm.summary.sn_low.x[, c("gene_id", "rsd", "avgrpkm")]
sn.go.df = sn.rsd.df
sn.go.df$gene_set = "IPSDSN-All genes"

terms = c("axonogenesis", "cell_differentiation", "cell_cycle",
          "CNS_development", "extracell_matrix_org",
          "nervous_system_dev", "neurogenesis",
          "sensory_organ_dev", "synapse_organization", "synaptic_signaling")

for (term in terms) {
  term.df = read.delim(paste0(outputDir, "/go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)
  term.df$gene_set = paste0("IPSDSN-", term)
  term.df = term.df %>% dplyr::left_join(sn.rsd.df) %>% na.omit()
  go.df2 = rbind(sn.go.df, term.df)
  go.df2.x = go.df2 %>% filter(avgrpkm > rpkm.threshold)
  p = ggplot(go.df2.x, aes(log(rsd), fill=gene_set)) +
    geom_density(alpha = 0.5) + theme_bw(fontSize) +
    xlab("log(CV)")
  print(p)
}
```

So the same is true among samples with low fibroblast content.
Let's compare with some GTEx nervous tissues to see whether these gene sets are highly variable in such a tissue, relative to other genes in that tissue.

```{r GOterms3}
rpkm.threshold = 1
tissue = "Brain - Cerebellum"
tissue = "Brain - Cortex"
tissue = "DRG"
tissue = "Nerve - Tibial"

tissue.rsd.df = rpkm.summary.df %>% filter(cell_type == tissue, avgrpkm > rpkm.threshold) %>% dplyr::select(gene_id, rsd, avgrpkm)
tissue.go.df = tissue.rsd.df
tissue.go.df$gene_set = paste0(tissue, "-All genes")

sn.rsd.df = rpkm.summary.df %>% filter(cell_type == "IPSDSN", avgrpkm > rpkm.threshold) %>% dplyr::select(gene_id, rsd, avgrpkm)
sn.go.df = sn.rsd.df
sn.go.df$gene_set = "IPSDSN-All genes"

allterms = c("axonogenesis", "cell_differentiation", "cell_cycle",
          "CNS_development", "extracell_matrix_org",
          "nervous_system_dev", "neurogenesis",
          "sensory_organ_dev", "synapse_organization", "synaptic_signaling")

terms = c("cell_cycle", "cell_differentiation", "extracell_matrix_org",
          "neurogenesis", "synapse_organization", "sensory_organ_dev")
#terms = c("axonogenesis")

# View density plots of CV across genes for all genes and specific GO terms,
# stratified by tissue
for (term in terms) {
  term.df = read.delim(paste0(outputDir, "/go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)
  term.df$gene_set = paste0(tissue, "-", term)
  term.df = term.df %>% dplyr::left_join(tissue.rsd.df) %>% na.omit()
  tissue.df2 = rbind(tissue.go.df, term.df) %>% filter(avgrpkm > rpkm.threshold)
  print(paste0(tissue, " median CV=", median(tissue.go.df$rsd)))
  print(paste0(tissue, " ", term, " CV=", median(term.df$rsd)))
  
  term.df = read.delim(paste0(outputDir, "/go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)
  term.df$gene_set = paste0("IPSDSN-", term)
  term.df = term.df %>% dplyr::left_join(sn.rsd.df) %>% na.omit()
  sn.df2 = rbind(sn.go.df, term.df) %>% filter(avgrpkm > rpkm.threshold)
  print(paste0("IPSDSN median CV=", median(sn.rsd.df$rsd)))
  print(paste0("IPSDSN ", term, " CV=", median(term.df$rsd)))
  
  tissue.df2$tissue = tissue
  sn.df2$tissue = "IPSDSN"
  go.df2.x = rbind(tissue.df2, sn.df2)
  p = ggplot(go.df2.x, aes(log(rsd), fill=tissue, color=gene_set)) +
    geom_density(alpha = 0.5) + theme_bw(fontSize) +
    xlab("log(CV)") + ggtitle(term)
  print(p)
}
```

Neuron-related gene sets all have elevated variability in IPSDSNs, whereas cell cycle genes do not. But this might not be the best way to visualise it. Let's try violin plots to see many tissues on the same plot.

```{r GOterms4, eval=FALSE, echo=FALSE}
# The first attempt, where I show a plot/panel for each GO term, comparing variability among genes in the GO term to all genes in the corresponding tissue.
log = T
printPlots = F
fontSize = 9
#tissues = c("Brain - Cerebellum", "DRG", "Nerve - Tibial", "IPSDSN")
tissues = c("Brain - Cerebellum", "DRG", "IPSC", "IPSDSN")
#terms = c("cell_cycle", "cell_differentiation", "extracell_matrix_org",
#          "neurogenesis", "synapse_organization", "sensory_organ_dev")
terms = c("cell_cycle", "apoptotic process",
          "cell_differentiation", "neurogenesis",
          "extracell_matrix_org", "sensory_organ_dev")

plots = list()
for (term in terms) {
  tissue.combined.df = data.frame()
  term.df = read.delim(paste0(outputDir, "/go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)
  tissueLevels = c()
  medians.df = data.frame()
  for (tissue in tissues) {
    tissue.allgenes = paste0(tissue, "-All genes")
    tissue.term = paste0(tissue, "-", term)
    
    tissue.rsd.df = rpkm.summary.df %>% filter(cell_type == tissue, avgrpkm > rpkm.threshold) %>% dplyr::select(gene_id, rsd, avgrpkm)
    tissue.go.df = tissue.rsd.df
    tissue.go.df$gene_set = tissue.allgenes
    term.df2 = term.df %>% dplyr::left_join(tissue.rsd.df) %>% na.omit()
    term.df2$gene_set = tissue.term
    tissue.df2 = rbind(tissue.go.df, term.df2) %>% filter(avgrpkm > rpkm.threshold)
    tissue.df2$tissue = tissue
    
    medians.df = rbind(medians.df, data.frame(gene_set=tissue.allgenes, med=median(tissue.go.df$rsd)))
    medians.df = rbind(medians.df, data.frame(gene_set=tissue.term, med=median(term.df2$rsd)))
    print(paste0(tissue, " median CV=", median(tissue.go.df$rsd)))
    print(paste0(tissue, "-", term, " CV=", median(term.df2$rsd)))
    
    tissue.combined.df = rbind(tissue.combined.df, tissue.df2)
    tissueLevels = c(tissueLevels, tissue.allgenes, tissue.term)
  }
  
  tissue.combined.df$gene_set = factor(as.character(tissue.combined.df$gene_set), levels=tissueLevels)
  medians.df$gene_set = factor(medians.df$gene_set, levels=tissueLevels)
  mediansText = formatC(medians.df$med, format="f", digits=3)
  termName = gsub("_", " ", term)
  if (log) {
    p = ggplot(tissue.combined.df, aes(x=gene_set, y=log(rsd), fill=tissue)) +
      geom_violin(alpha=0.6) + geom_boxplot(width=0.1, outlier.shape = NA) +
      theme_bw(fontSize) + theme(legend.position="None") +
      theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
      ggtitle(termName) + ylab("log(CV)") +
      annotate("text", x=medians.df$gene_set, y=2.2, label=mediansText, size=2.5)
  } else {
    p = ggplot(tissue.combined.df, aes(x=gene_set, y=rsd, fill=tissue)) +
      geom_violin(alpha=0.6) + geom_boxplot(width=0.1, outlier.shape = NA) +
      theme_bw(fontSize) + theme(legend.position="None") +
      theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
      ggtitle(termName) + ylab("CV") +
      coord_cartesian(ylim=c(0, 2)) +
      annotate("text", x=medians.df$gene_set, y=1.8, label=mediansText, size=2.5)
  }
  if (printPlots) {
    print(p)
  }
  plots = c(plots, list(p + theme(axis.text.x = element_blank())))
}

grid.arrange(grobs=plots, ncol=2)

# Make a plot to get the x axis labels
IPSDSN.df = tissue.combined.df %>% filter(tissue == "IPSDSN")
IPSDSN.df$gene_set = gsub("IPSDSN-", "", IPSDSN.df$gene_set)
IPSDSN.df$gene_set[IPSDSN.df$gene_set != "All genes"] = "GO subset"
p = ggplot(IPSDSN.df, aes(x=gene_set, y=log(rsd), fill=tissue)) +
  geom_violin(alpha=0.6) + geom_boxplot(width=0.1, outlier.shape = NA) +
  theme_bw(fontSize) + theme(legend.position="None") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
  ggtitle(termName) + ylab("log(CV)")
print(p)
```


```{r GOterms5}
# The second attempt, based on Dan's suggestion, where I show plots by tissue.
# This way I only need to include the baseline "all genes" violin once per tissue,
# and it's easy to see whether neuronal GO terms are more highly variable in general
# than non-neuronal GO terms.
log = T
printPlots = T
fontSize = 10
tissues = c("Brain - Cerebellum", "DRG", "Nerve - Tibial", "IPSC", "IPSDSN")
tissues = c("Brain - Cerebellum", "IPSC")
tissues = c("IPSDSN")

terms.df = data.frame(term = c("cell_differentiation", "neurogenesis", "synapse_organization", "sensory_organ_dev",
                               "cell_cycle", "vesicle membrane", "apoptotic process",
                               "immune response", "receptor binding", 
                               "extracell_matrix_org"),
                      category = c("neural", "neural", "neural", "neural", "general",
                                   "general", "general", "general", "general", "general"))

plots = list()
for (tissue in tissues) {
  tissue.combined.df = data.frame()
  medians.df = data.frame()
  termName = "All genes"
  termLevels = c(termName)
  tissue.rsd.df = rpkm.summary.df %>% filter(cell_type == tissue, avgrpkm > rpkm.threshold) %>% dplyr::select(gene_id, rsd, avgrpkm)
  tissue.go.df = tissue.rsd.df
  tissue.go.df$gene_set = termName
  tissue.go.df$category = "All genes"
  
  medians.df = rbind(medians.df, data.frame(gene_set=termName, med=median(tissue.go.df$rsd)))
  tissue.combined.df = rbind(tissue.combined.df, tissue.go.df)
  
  for (i in 1:length(terms.df$term)) {
  #for (term in terms.df$term) {
    term = terms.df$term[i]
    category = terms.df$category[i]
    termName = gsub("_", " ", term)
    term.df = read.delim(paste0(outputDir, "/go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)

    term.df = term.df %>% dplyr::left_join(tissue.rsd.df) %>% na.omit()
    term.df$gene_set = termName
    term.df$category = category

    medians.df = rbind(medians.df, data.frame(gene_set=termName, med=median(term.df$rsd)))
    print(paste0(tissue, "-", term, " CV=", median(term.df$rsd)))
    
    tissue.combined.df = rbind(tissue.combined.df, term.df)
    termLevels = c(termLevels, termName)
  }
  
  tissue.combined.df$gene_set = factor(as.character(tissue.combined.df$gene_set), levels=termLevels)
  medians.df$gene_set = factor(medians.df$gene_set, levels=termLevels)
  mediansText = formatC(medians.df$med, format="f", digits=3)
  
  if (log) {
    p = ggplot(tissue.combined.df, aes(x=gene_set, y=log(rsd), fill=category)) +
      geom_violin(alpha=0.6) + geom_boxplot(width=0.1, outlier.shape = NA) +
      theme_bw(fontSize) + theme(legend.position="None") +
      theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
      ggtitle(tissue) + ylab("log(CV)") +
      annotate("text", x=medians.df$gene_set, y=2.2, label=mediansText, size=3)
  } else {
    p = ggplot(tissue.combined.df, aes(x=gene_set, y=rsd, fill=category)) +
      geom_violin(alpha=0.6) + geom_boxplot(width=0.1, outlier.shape = NA) +
      theme_bw(fontSize) + theme(legend.position="None") +
      theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
      ggtitle(tissue) + ylab("CV") +
      coord_cartesian(ylim=c(0, 2)) +
      annotate("text", x=medians.df$gene_set, y=1.8, label=mediansText, size=3)
  }
  if (printPlots) {
    print(p)
  }
  if (tissue == "IPSDSN") {
    plots = c(plots, list(p))
  } else {
    plots = c(plots, list(p + theme(axis.text.x = element_blank())))
  }
}

layout <- rbind(c(1), c(1),
                c(2), c(2),
                c(3), c(3),
                c(4), c(4),
                c(5), c(5), c(5))
grid.arrange(grobs=plots, layout_matrix=layout)
#grid.arrange(grobs=plots, ncol=1)

# Make a plot to get the x axis labels
IPSDSN.df = tissue.combined.df %>% filter(tissue == "IPSDSN")
IPSDSN.df$gene_set = gsub("IPSDSN-", "", IPSDSN.df$gene_set)
IPSDSN.df$gene_set[IPSDSN.df$gene_set != "All genes"] = "GO subset"
p = ggplot(IPSDSN.df, aes(x=gene_set, y=log(rsd), fill=tissue)) +
  geom_violin(alpha=0.6) + geom_boxplot(width=0.1, outlier.shape = NA) +
  theme_bw(fontSize) + theme(legend.position="None") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
  ggtitle(termName) + ylab("log(CV)")
print(p)
```


```{r GOterms6}
# Another version, where we plot the median for all genes and TERM genes
# as a point with line between them
log = T
printPlots = T
fontSize = 9
#tissues = c("Brain - Cerebellum", "DRG", "Nerve - Tibial", "IPSDSN")
tissues = c("Brain - Cerebellum", "DRG", "IPSC", "Nerve - Tibial", "IPSDSN")
terms = c("cell_cycle", "apoptotic process", "vesicle membrane",
             "immune response",
             "cell_differentiation", "CNS_development", "nervous_system_dev",
             "extracell_matrix_org", "axonogenesis", "neurogenesis", 
             "sensory_organ_dev", "synaptic_signaling")
termNames = gsub("_", " ", terms)

all.medians.df = data.frame()
for (term in terms) {
  print(term)
  term.df = read.delim(paste0(outputDir, "go/go.", term, ".genes.txt")) %>% dplyr::select(gene_id = Gene.ID)
  termLevels = c("All genes", "GO term")

  term.medians.df = data.frame()
  for (tissue in tissues) {
    tissue.rsd.df = rpkm.summary.df %>% filter(cell_type == tissue, avgrpkm > rpkm.threshold) %>% dplyr::select(gene_id, rsd, avgrpkm)
    tissue.go.df = tissue.rsd.df
    tissue.go.df$gene_set = "All genes"
    term.df2 = term.df %>% dplyr::left_join(tissue.rsd.df) %>% na.omit()
    term.df2$gene_set = term

    term.medians.df = rbind(term.medians.df, data.frame(gene_set="All genes", tissue=tissue, median.rsd=median(tissue.go.df$rsd)))
    term.medians.df = rbind(term.medians.df, data.frame(gene_set="GO term", tissue=tissue, median.rsd=median(term.df2$rsd)))
    print(paste0(tissue, " median CV=", median(tissue.go.df$rsd)))
    print(paste0(tissue, "-", term, " CV=", median(term.df2$rsd)))
  }
  term.medians.df$gene_set = factor(term.medians.df$gene_set, levels=termLevels)
  term.medians.df$termName = gsub("_", " ", term)
  all.medians.df = rbind(all.medians.df, term.medians.df)
}

all.medians.df = all.medians.df %>%
  mutate(tissue=ifelse(tissue == "Brain - Cerebellum", "Cerebellum", tissue)) %>%
  mutate(tissue=ifelse(tissue == "Nerve - Tibial", "Tibial nerve", tissue))
all.medians.df$termName = factor(as.character(all.medians.df$termName), levels=termNames)
all.medians.df$tissue = factor(as.character(all.medians.df$tissue), levels=c("Cerebellum", "DRG", "IPSC", "Tibial nerve", "IPSDSN"))
axisScales = "free_y"

ggplot(all.medians.df, aes(x=gene_set, y=median.rsd, col=tissue, group=tissue)) +
  geom_line(size=2, alpha=0.8) + geom_point(alpha=0.7, size=4) +
  theme_bw(9) +
  theme(axis.title.x = element_blank()) +
  ylab("log(median CV)") +
  scale_x_discrete(expand=c(0.2,0.2)) +
  facet_wrap(~termName, ncol = 3, scale = axisScales) +
  theme(strip.background = element_rect(fill="white", linetype="blank"), strip.text = element_text(size=9)) +
  theme(axis.text.x = element_text(size=8), panel.margin.y=unit(0.3, "cm")) +
  scale_color_discrete(name="Tissue")

subsetTerms = c("cell cycle", "vesicle membrane", "immune response",
             "nervous system dev", "extracell matrix org", "sensory organ dev")
subset.medians.df = all.medians.df %>% filter(termName %in% subsetTerms)
ggplot(subset.medians.df, aes(x=gene_set, y=median.rsd, col=tissue, group=tissue)) +
  geom_line(size=2, alpha=0.8) + geom_point(alpha=0.7, size=4) +
  theme_bw(10) +
  theme(axis.title.x = element_blank()) +
  ylab("log(median CV)") +
  scale_x_discrete(expand=c(0.2,0.2)) +
  facet_wrap(~termName, ncol = 3, scale = axisScales) +
  theme(strip.background = element_rect(fill="white", linetype="blank"), strip.text = element_text(size=10)) +
  theme(axis.text.x = element_text(size=9), panel.margin.y=unit(0.3, "cm")) +
  scale_color_discrete(name="Tissue")
```

Let's also see the effect of sample RIN number.

```{r CalculateRSD_RIN_Subsets}
sn.meta.P2 <- sn.meta.all %>% filter(Protocol == "P2V2") %>% mutate(sampleDifferentiation=paste(donor, Differentiaton.Replicate))

ggplot(sn.meta.P2, aes(x=RIN)) + geom_histogram() +
  theme_bw(18) + ggtitle("IPSDSN sample RNA integrity number")
mean(sn.meta.P2$RIN, na.rm=T)

rpkm.df.sn = sn.df.all %>% dplyr::select(-length) %>% filter(gene_id %in% annotation$gene_id)
rownames(rpkm.df.sn) = rpkm.df.sn$gene_id

highRINsampleIDs = sn.meta.P2[!is.na(sn.meta.P2$RIN) & sn.meta.P2$RIN >= 8, ]$Sample
rpkm.mat.highRIN = as.matrix(rpkm.df.sn[ , highRINsampleIDs])
rownames(rpkm.mat.highRIN) = rpkm.df.sn$gene_id
rpkm.summary.highRIN.x = getRSDSummary(rpkm.mat.highRIN, "IPSDSN-highRIN")

rpkm.mat.lowRIN = as.matrix(rpkm.df.sn[ , sn.meta.P2[!is.na(sn.meta.P2$RIN) & sn.meta.P2$RIN < 8, ]$Sample])
rownames(rpkm.mat.lowRIN) = rpkm.df.sn$gene_id
rpkm.summary.lowRIN.x = getRSDSummary(rpkm.mat.lowRIN, "IPSDSN-lowRIN")

# Compare the difference between high/low RIN samples with DRG.
rpkm.summary.RINsubsets = rbind(rpkm.summary.lowRIN.x,
                                rpkm.summary.highRIN.x,
                                rpkm.summary.sn.x,
                                rpkm.summary.df[rpkm.summary.df$cell_type == "DRG",])
rpkm.summary.RINsubsets$cell_type[rpkm.summary.RINsubsets$cell_type == "IPSDSN"] = "IPDSDN-all"
p = ggplot(rpkm.summary.RINsubsets, aes(log(rsd), fill=cell_type)) +
  geom_density(alpha = 0.5) + theme_bw(fontSize) +
  theme(legend.position=c(1, 1), legend.justification=c(1,1)) +
  xlab("log(CV)")
print(p)

quantile(rpkm.summary.lowRIN.x$rsd, probs=c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99))
quantile(rpkm.summary.highRIN.x$rsd, probs=c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99))
quantile(rpkm.summary.sn.x$rsd, probs=c(0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99))
```

Check the effect of sample size on the RSD distribution. I'll take the largest GTEx sample, skeletal muscle (430 samples), and subsample it. For smaller subsamples, I do this repeatedly to average the RSD distributions of the subsamples. E.g. when subsampling to 10% of the total samples, I do this and calculate the RSDs for all genes 10 times, and paste the results together, i.e. we'll have 10 numbers for every gene, one from each of the 10 times I subsampled. This just makes the distributions smoother and easier to compare.

```{r RSD_sample_size}
# Number of muscle samples:
sum(sample.meta$SMTSD == "Muscle - Skeletal")

rpkm.summary.ss.1 = plotTissueRSDs(c("Muscle - Skeletal"), makePlots=F, subsample=1) %>%
  filter(cell_type == "Muscle - Skeletal")

rpkm.summary.ss.0.5 = data.frame()
for (i in 1:2) {
  rpkm.summary.ss.0.5 = rbind(rpkm.summary.ss.0.5,
        plotTissueRSDs(c("Muscle - Skeletal"), makePlots=F, subsample=0.5) %>%
          filter(cell_type == "Muscle - Skeletal") %>%
          mutate(cell_type = "Muscle - 0.5") )
}

rpkm.summary.ss.0.25 = data.frame()
for (i in 1:4) {
  rpkm.summary.ss.0.25 = rbind(rpkm.summary.ss.0.25,
        plotTissueRSDs(c("Muscle - Skeletal"), makePlots=F, subsample=0.25) %>%
          filter(cell_type == "Muscle - Skeletal") %>%
          mutate(cell_type = "Muscle - 0.25") )
}

rpkm.summary.ss.0.1 = data.frame()
for (i in 1:10) {
  rpkm.summary.ss.0.1 = rbind(rpkm.summary.ss.0.1,
        plotTissueRSDs(c("Muscle - Skeletal"), makePlots=F, subsample=0.1) %>%
          filter(cell_type == "Muscle - Skeletal") %>%
          mutate(cell_type = "Muscle - 0.1") )
}

rpkm.summary.ss.0.05 = data.frame()
for (i in 1:20) {
  rpkm.summary.ss.0.05 = rbind(rpkm.summary.ss.0.05,
        plotTissueRSDs(c("Muscle - Skeletal"), makePlots=F, subsample=0.05) %>%
          filter(cell_type == "Muscle - Skeletal") %>%
          mutate(cell_type = "Muscle - 0.05") )
}

rpkm.summary.ss = rbind(rpkm.summary.ss.1, rpkm.summary.ss.0.5, rpkm.summary.ss.0.25, rpkm.summary.ss.0.1, rpkm.summary.ss.0.05)
rpkm.summary.ss$Subset = rpkm.summary.ss$cell_type
rpkm.summary.ss$Subset[rpkm.summary.ss$Subset == "Muscle - Skeletal"] = "Muscle - all"
ggplot(rpkm.summary.ss, aes(log(rsd), fill=Subset)) + geom_density(alpha = 0.5) + theme_bw(fontSize) + xlab("log(CV)")
```

There is clearly no effect of sample size on the RSD distribution.

Let's now see what comes up as enriched considering the high RSD genes in some other tissues.
The order is: DRG, IPSC, Tibial nerve, Fibroblasts, Whole blood.

```{r GOenrichment2}
tmp.rpkm.df = rpkm.summary.df[rpkm.summary.df$cell_type == "DRG",]
tmp.rpkm.df = arrange(tmp.rpkm.df, -rsd)
gl = list(DRG=tmp.rpkm.df[1:1000,]$gene_id)
gs = gosummaries(gl, max_signif=100)
plot(gs, fontsize = goSummariesFontSize)
gs.results.df = attributes(gs)$gprofiler_results
fname = paste0(outputDir, "highRSDgenes.DRG.gosummaries.results.txt")
write.table(gs.results.df[,4:15], fname, quote=F, sep="\t", row.names=F)

tmp.rpkm.df = rpkm.summary.df[rpkm.summary.df$cell_type == "IPSC",]
tmp.rpkm.df = arrange(tmp.rpkm.df, -rsd)
gl = list(IPSC=tmp.rpkm.df[1:1000,]$gene_id)
gs = gosummaries(gl, max_signif=100)
plot(gs, fontsize = goSummariesFontSize)
gs.results.df = attributes(gs)$gprofiler_results
fname = paste0(outputDir, "highRSDgenes.IPSC.gosummaries.results.txt")
write.table(gs.results.df[,4:15], fname, quote=F, sep="\t", row.names=F)

tmp.rpkm.df = rpkm.summary.df[rpkm.summary.df$cell_type == "Nerve - Tibial",]
tmp.rpkm.df = arrange(tmp.rpkm.df, -rsd)
gl = list(TibialNerve=tmp.rpkm.df[1:1000,]$gene_id)
gs = gosummaries(gl, max_signif=100)
plot(gs, fontsize = goSummariesFontSize)
gs.results.df = attributes(gs)$gprofiler_results
fname = paste0(outputDir, "highRSDgenes.Nerve.gosummaries.results.txt")
write.table(gs.results.df[,4:15], fname, quote=F, sep="\t", row.names=F)

tmp.rpkm.df = rpkm.summary.df[rpkm.summary.df$cell_type == "Fibroblasts",]
tmp.rpkm.df = arrange(tmp.rpkm.df, -rsd)
gl = list(Fibroblasts=tmp.rpkm.df[1:1000,]$gene_id)
gs = gosummaries(gl, max_signif=100)
plot(gs, fontsize = goSummariesFontSize)
gs.results.df = attributes(gs)$gprofiler_results
fname = paste0(outputDir, "highRSDgenes.Fibroblasts.gosummaries.results.txt")
write.table(gs.results.df[,4:15], fname, quote=F, sep="\t", row.names=F)

tmp.rpkm.df = rpkm.summary.df[rpkm.summary.df$cell_type == "Whole Blood",]
tmp.rpkm.df = arrange(tmp.rpkm.df, -rsd)
gl = list(WholeBlood=tmp.rpkm.df[1:1000,]$gene_id)
gs = gosummaries(gl, max_signif=100)
plot(gs, fontsize = goSummariesFontSize)
gs.results.df = attributes(gs)$gprofiler_results
fname = paste0(outputDir, "highRSDgenes.Blood.gosummaries.results.txt")
write.table(gs.results.df[,4:15], fname, quote=F, sep="\t", row.names=F)
```

We are interested not just in genes with high RSD across samples, but those genes where the RSD is high in SNs and NOT in other tissues. I.e. a relative RSD. To determine this, I take a gene's RSD in the SNs and divide it by the mean RSD across other tissues (considering only tissues where that gene passes the expression threshold), to get an RSDratio per gene. And I do the same in other individual tissues to see what "highly variable" genes are found in different tissues.

```{r RelativeRSD}
rpkm.summary.spread = rpkm.summary.df %>% dplyr::select(-avgrpkm, -logavgrpkm) %>% tidyr::spread(cell_type, rsd)
rpkm.summary.spread$tissueRSDMean = apply(dplyr::select(rpkm.summary.spread, -IPSDSN, -gene_id), 1, mean_na_rm)
rpkm.summary.spread$snRSDRatio = rpkm.summary.spread$IPSDSN / rpkm.summary.spread$tissueRSDMean

rpkm.summary.RSDRatio = data.frame(gene_id = rpkm.summary.spread$gene_id)
#tissueNames = colnames(select(rpkm.summary.spread, -IPSDSN, -gene_id))
for (tissue in tissueNames) {
  tissueRSDMean = apply(dplyr::select(rpkm.summary.spread, -IPSDSN, -gene_id, -matches(tissue)), 1, mean_na_rm)
  rpkm.summary.RSDRatio[, tissue] = rpkm.summary.spread[, tissue] / tissueRSDMean
}
rpkm.summary.RSDRatio[, "IPSDSN"] = rpkm.summary.spread$IPSDSN / rpkm.summary.spread$tissueRSDMean

rpkm.summary.RSDRatio.gather = gather(rpkm.summary.RSDRatio, cell_type, RSDRatio, 2:ncol(rpkm.summary.RSDRatio))
if (!knitting) {
  ggplot(rpkm.summary.RSDRatio.gather, aes(log(RSDRatio), fill=cell_type)) + geom_density(alpha = 0.5) + theme_bw(9)
}

# Add RSDRatio to the table with RSD and RPKM per gene for IPSDSNs,
# and save it as a table.
rpkm.summary.sn.x.ann = rpkm.summary.sn.x %>% dplyr::left_join(annotation, by="gene_id") %>%
  left_join(rpkm.summary.RSDRatio[, c("gene_id", "IPSDSN")]) %>%
  rename(rsdRatio = IPSDSN)
fname = paste0(outputDir, "rpkm.summary.sn.rsdratio.txt")
write.table(rpkm.summary.sn.x.ann, fname, quote=F, sep="\t", row.names=F)

highlightedTissues = c("Whole Blood", "IPSC", "DRG", "IPSDSN")
# Set only a couple of the cell types to be coloured
tn = factor(c(tissueNames[!tissueNames %in% highlightedTissues], highlightedTissues),
            levels = c(tissueNames[!tissueNames %in% highlightedTissues], highlightedTissues))
rpkm.summary.RSDRatio.gather$cell_type = factor(rpkm.summary.RSDRatio.gather$cell_type, levels=levels(tn))
p = ggplot(rpkm.summary.RSDRatio.gather, aes(log(RSDRatio), fill=cell_type)) + geom_density() + theme_bw(fontSize)
p = p + scale_fill_manual(breaks = tn,
                          values=c(rep(rgb(0.5, 0.5, 0.5, alpha=0.5), length(tissueNames)-3), rgb(0, 0, 1, alpha=0.5), rgb(0.1, 0.7, 0.1, alpha=0.5), rgb(1, 0.65, 0, alpha=0.5), rgb(1, 0, 0, alpha=0.7))) + xlim(-2,2) +
  theme(legend.position="none") + xlab("log(CVRatio)")
print(p)

# This next plot is just to get the legend, which merges "other" GTEx tissues together,
# and which we'll paste onto the figure just above
rpkm.summary.RSDRatio.gather.df2 = rpkm.summary.RSDRatio.gather
rpkm.summary.RSDRatio.gather.df2$cell_type = as.character(rpkm.summary.RSDRatio.gather.df2$cell_type)
rpkm.summary.RSDRatio.gather.df2$cell_type[!(rpkm.summary.RSDRatio.gather.df2$cell_type %in% highlightedTissues)] = "GTEx-other"
rpkm.summary.RSDRatio.gather.df2$cell_type = factor(rpkm.summary.RSDRatio.gather.df2$cell_type, levels=c("GTEx-other", highlightedTissues))
p = ggplot(rpkm.summary.RSDRatio.gather.df2, aes(log(RSDRatio), fill=cell_type)) + geom_density() + theme_bw(fontSize) + theme(legend.title=element_blank())
p = p + scale_fill_manual(breaks = levels(rpkm.summary.RSDRatio.gather.df2$cell_type),
                          values=c(rgb(0.5, 0.5, 0.5, alpha=0.5), rgb(0, 0, 1, alpha=0.5), rgb(0.1, 0.7, 0.1, alpha=0.5), rgb(1, 0.65, 0, alpha=0.5), rgb(1, 0, 0, alpha=0.7))) + xlim(-2,2) + xlab("log(CVRatio)")
print(p)



q = quantile(rpkm.summary.RSDRatio.gather %>% filter(cell_type != "IPSDSN") %>% .[["RSDRatio"]],
             probs = qprobs, na.rm=T)
q
# 99% of genes in all tissue samples have RSD < 2.21, but this differs by tissue
sum(rpkm.summary.RSDRatio$IPSDSN > q["99%"], na.rm = T)
sum(rpkm.summary.RSDRatio$`Brain - Cerebellum` > q["99%"], na.rm = T)
sum(rpkm.summary.RSDRatio$DRG > q["99%"], na.rm = T)
sum(rpkm.summary.RSDRatio$Nerve > q["99%"], na.rm = T)
sum(rpkm.summary.RSDRatio$`Whole Blood` > q["99%"], na.rm = T)

#quantile(rpkm.summary.RSDRatio$IPSDSN, probs = qprobs, na.rm=T)
rsdRatio.quantiles.df = data.frame()
for (tissue in c("IPSDSN", tissueNames)) {
  q = quantile(rpkm.summary.RSDRatio.gather$RSDRatio[rpkm.summary.RSDRatio.gather$cell_type == tissue], probs = c(0.5, 0.9, 0.95, 0.99), na.rm=T)
  avg = mean_na_rm(rpkm.summary.RSDRatio.gather$RSDRatio[rpkm.summary.RSDRatio.gather$cell_type == tissue])
  rsdRatio.quantiles.df = rbind(rsdRatio.quantiles.df, c(cell_type = tissue, avg, q))
}
colnames(rsdRatio.quantiles.df) = c("cell_type", "mean", "50%", "90%", "95%", "99%")
```

Looking at RSDratio shows the SNs (and whole blood) to be a larger outlier than considering RSD alone. But overall the result is very similar, so it doesn't seem worth the extra complication.

Let's see if there's a difference in how many eQTLs called by RASQUAL and FastQTL specifically among the genes with high RSD.
First, we note that FastQTL calls more QTLs at lowly expressed genes (RPKM < 1). 45% of FastQTLs eGenes have RPKM < 1. 

```{r RasqualFastQTLHighRSDGenes}
fastQTL.df = read.delim("results/fastqtl/permutations.10k.allgenes.cis500k.PCs.20.fdr0.1.txt", sep=" ")
rasqual.df = read.delim("results/rasqual/nopu.rasqual.500k.eigenMT.97samples.merged.leadSNPs.fdr0.1.txt")
rasqual.df = rasqual.df %>% dplyr::rename(geneid = gene)

rpkm.summary.sn = rpkm.summary.sn[order(rpkm.summary.sn$rsd),]
rpkm.summary.sn.x = rpkm.summary.sn[rpkm.summary.sn$avgrpkm > rpkmThreshold,]

rasqual.df = dplyr::inner_join(rasqual.df, rpkm.summary.sn, by=c("geneid"="gene_id"))
fastQTL.df = dplyr::inner_join(fastQTL.df, rpkm.summary.sn, by=c("geneid"="gene_id"))
rownames(fastQTL.df) = fastQTL.df$geneid
rownames(rasqual.df) = rasqual.df$geneid

# Number of EXPRESSED eGenes for each method
sum(fastQTL.df$avgrpkm > rpkmThreshold)
sum(rasqual.df$avgrpkm > rpkmThreshold)

plot.df = rbind(data.frame(gene_set="All genes", rsd=rpkm.summary.sn$rsd, logavgrpkm=log(rpkm.summary.sn$avgrpkm, base=10)),
                data.frame(gene_set="RASQUAL eGenes", rsd=rasqual.df$rsd, logavgrpkm=log(rasqual.df$avgrpkm, base=10)),
                data.frame(gene_set="FastQTL eGenes", rsd=fastQTL.df$rsd, logavgrpkm=log(fastQTL.df$avgrpkm, base=10)))

ggplot(plot.df[plot.df$gene_set != "All genes",], aes(logavgrpkm, fill=gene_set)) + geom_histogram(alpha=0.5, position="dodge") + theme_bw(fontSize)

ggplot(plot.df, aes(logavgrpkm, fill=gene_set)) + geom_density(alpha=0.5) + theme_bw(fontSize)
```

Because lowly expressed genes have higher relative variability, we need to stratify by RPKM when we look at how many eGenes we get with FastQTL and RASQUAL at different RSD.
Let's look at the number of Rasqual eGenes and FastQTL eGenes in different RSD bins. We can also compare these with the density of all genes in each RSD bin.

```{r RasqualFastQTLHighRSDGenes2}
plot.df$category=NA
plot.df$category[plot.df$logavgrpkm < 0] = "RPKM < 1"
plot.df$category[plot.df$logavgrpkm > 0 & plot.df$logavgrpkm < 1] = "RPKM 1 - 10"
plot.df$category[plot.df$logavgrpkm > 1] = "RPKM > 10"
plot.df$category = factor(plot.df$category, levels=c("RPKM < 1", "RPKM 1 - 10", "RPKM > 10"))

ggplot(plot.df[plot.df$gene_set != "All genes",], aes(log(rsd), fill=gene_set)) +
  geom_histogram(alpha=0.5, position="dodge") +
  theme_bw(fontSize) +
  facet_wrap(~category, ncol=1) +
  xlab("log(CV)")
ggplot(plot.df[plot.df$gene_set != "All genes",], aes(log(rsd), fill=gene_set)) +
  geom_density(alpha=0.5) +
  theme_bw(fontSize) +
  facet_wrap(~category, ncol=1) +
  xlab("log(CV)")

# Also compare the densities to All genes, separately for Rasqual and FastQTL
ggplot(plot.df[plot.df$gene_set != "RASQUAL eGenes",], aes(log(rsd), fill=gene_set)) +
  geom_density(alpha=0.5) +
  theme_bw(fontSize) +
  facet_wrap(~category, ncol=1) +
  xlab("log(CV)")
ggplot(plot.df[plot.df$gene_set != "FastQTL eGenes",], aes(log(rsd), fill=gene_set)) +
  geom_density(alpha=0.5) +
  theme_bw(fontSize) +
  facet_wrap(~category, ncol=1) +
  xlab("log(CV)")
```

We see above that Rasqual calls more eQTLs for genes with high RSD, except for very lowly expressed genes. For genes with very low expression FastQTL calls more eQTLs... but this isn't necessarily a good thing.

We should check whether the rate of "novel" eQTLs, i.e. those not in GTEx, changes across RSD bins.

```{r RasqualFastQTLHighRSDGenes3}
# Add info on whether each eQTL is known or novel
rasqual.tissuespecific = read.delim("results/tissuespecific/Neuron_Sensory_rasqual.leadVar.ld_defined_tissue_specific.txt")
fastqtl.tissuespecific = read.delim("results/tissuespecific/Neuron_Sensory_fastqtl.leadVar.ld_defined_tissue_specific.txt")

# First check how many eGenes discovered by FastQTL were discovered by RASQUAL
# and vice versa
fastQTL.in.rasqual = sum(fastQTL.df$geneid %in% rasqual.df$geneid)

getFastQTLFractionFound = function(fdr) {
  df = fastQTL.df %>% filter(FDR < fdr)
  sum(df$geneid %in% rasqual.df$geneid) / dim(df)[1]
}
writeLines("Fraction of fastQTL eGenes (at different FDR thresholds) found in Rasqual\n0.1\t0.01\t0.001\t0.0001")
sapply(c(0.1, 0.01, 0.001, 0.0001), getFastQTLFractionFound)

getRasqualFractionFound = function(fdr) {
  df = rasqual.df %>% filter(FDR < fdr)
  sum(df$geneid %in% fastQTL.df$geneid) / dim(df)[1]
}
writeLines("Fraction of RASQUAL eGenes (at different FDR thresholds) found in FastQTL\n0.1\t0.01\t0.001\t0.0001")
sapply(c(0.1, 0.01, 0.001, 0.0001), getRasqualFractionFound)
# RASQUAL FDR can't be below 0.001, since we did only 1000 permutations

rasqual.df$eqtl_status = "known"
rasqual.df[rasqual.tissuespecific$ensid,]$eqtl_status = "novel"
fastQTL.df$eqtl_status = "known"
fastQTL.df[fastqtl.tissuespecific$ensid,]$eqtl_status = "novel"

# Consider only eGenes in either FastQTL or Rasqual, and get the quantiles of RSD.
egene.df = rbind(data.frame(rasqual.df[, c("geneid", "rsd", "avgrpkm", "eqtl_status")], method="RASQUAL"),
                 data.frame(fastQTL.df[, c("geneid", "rsd", "avgrpkm", "eqtl_status")], method="FastQTL"))
egene.df = na.omit(egene.df)
#egene.df = egene.df[!duplicated(egene.df$geneid),]

numQuantiles = 4

egene.df = egene.df[egene.df$avgrpkm >= rpkmThreshold,]
rsd.quantiles = quantile(egene.df[!duplicated(egene.df$geneid),]$rsd, seq(0, 1, 1/numQuantiles), na.rm=T)

egene.df$quantile = NA
egene.df$quantile[egene.df$rsd >= rsd.quantiles[1] & egene.df$rsd <= rsd.quantiles[2]] = "0 - 0.25"
egene.df$quantile[egene.df$rsd >= rsd.quantiles[2] & egene.df$rsd <= rsd.quantiles[3]] = "0.25 - 0.5"
egene.df$quantile[egene.df$rsd >= rsd.quantiles[3] & egene.df$rsd <= rsd.quantiles[4]] = "0.5 - 0.75"
egene.df$quantile[egene.df$rsd >= rsd.quantiles[4] & egene.df$rsd <= rsd.quantiles[5]] = "0.75 - 1.0"
egene.df$quantile = factor(egene.df$quantile, levels=c("0 - 0.25", "0.25 - 0.5", "0.5 - 0.75", "0.75 - 1.0"))

ggplot(egene.df, aes(x=method, fill=eqtl_status)) +
  geom_bar() +
  theme_bw(fontSize) +
  facet_wrap(~quantile, nrow=1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank()) +
  ggtitle("Number of eGenes across quartiles of gene variability")


fastqtl.counts.novel = sapply(1:numQuantiles, FUN=function(x) {
  df = fastQTL.df[fastQTL.df$avgrpkm >= rpkmThreshold & fastQTL.df$eqtl_status == "novel",]
  sum(df$rsd >= rsd.quantiles[x] & df$rsd <= rsd.quantiles[x+1], na.rm=T)
})
rasqual.counts.novel = sapply(1:numQuantiles, FUN=function(x) {
  df = rasqual.df[rasqual.df$avgrpkm >= rpkmThreshold & rasqual.df$eqtl_status == "novel",]
  sum(df$rsd >= rsd.quantiles[x] & df$rsd <= rsd.quantiles[x+1], na.rm=T)
})
fastqtl.counts.known = sapply(1:numQuantiles, FUN=function(x) {
  df = fastQTL.df[fastQTL.df$avgrpkm >= rpkmThreshold & fastQTL.df$eqtl_status == "known",]
  sum(df$rsd >= rsd.quantiles[x] & df$rsd <= rsd.quantiles[x+1], na.rm=T)
})
rasqual.counts.known = sapply(1:numQuantiles, FUN=function(x) {
  df = rasqual.df[rasqual.df$avgrpkm >= rpkmThreshold & rasqual.df$eqtl_status == "known",]
  sum(df$rsd >= rsd.quantiles[x] & df$rsd <= rsd.quantiles[x+1], na.rm=T)
})
fastqtl.counts = sapply(1:numQuantiles, FUN=function(x) {
  df = fastQTL.df[fastQTL.df$avgrpkm >= rpkmThreshold,]
  sum(df$rsd >= rsd.quantiles[x] & df$rsd <= rsd.quantiles[x+1], na.rm=T)
})
rasqual.counts = sapply(1:numQuantiles, FUN=function(x) {
  df = rasqual.df[rasqual.df$avgrpkm >= rpkmThreshold,]
  sum(df$rsd >= rsd.quantiles[x] & df$rsd <= rsd.quantiles[x+1], na.rm=T)
})
count.ratio = rasqual.counts / fastqtl.counts
count.ratio

known.ratio = rasqual.counts.known / fastqtl.counts.known
known.ratio

novel.ratio = data.frame(method="FastQTL", quantile=levels(egene.df$quantile), novel_ratio=(fastqtl.counts.novel / fastqtl.counts))
novel.ratio = rbind(novel.ratio, data.frame(method="RASQUAL", quantile=levels(egene.df$quantile), novel_ratio=(rasqual.counts.novel / rasqual.counts)))
ggplot(novel.ratio, aes(x=method, y=novel_ratio)) +
  geom_bar(stat="identity") +
  theme_bw(fontSize) +
  facet_wrap(~quantile, nrow=1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank()) +
  ggtitle("Fraction of novel eQTLs")

qtl.counts.df = rbind(data.frame(quantile=c("0 - 0.25", "0.25 - 0.5", "0.5 - 0.75", "0.75 - 1.0"), count=fastqtl.counts, method="fastQTL"),
                      data.frame(quantile=c("0 - 0.25", "0.25 - 0.5", "0.5 - 0.75", "0.75 - 1.0"), count=rasqual.counts, method="RASQUAL"))

ggplot(qtl.counts.df, aes(x=quantile, y=count, fill=method)) + geom_bar(stat = "identity", position="dodge") + theme_bw(fontSize)

ggplot(data.frame(ratio=known.ratio, quantile=c("0 - 0.25", "0.25 - 0.5", "0.5 - 0.75", "0.75 - 1.0")), aes(x=quantile)) +
  geom_bar(aes(y=ratio), stat = "identity") +
  theme_bw(fontSize) +
  labs(title="Ratio of RASQUAL eQTLs to FastQTL\n(known eQTLs only)", x="Quantile of gene expression variability")

```

